wallet/src/actors/app/methods.rs

```diff
@@ -818,4 +829,147 @@ impl App {
 
         Box::new(f)
     }
+
+    /// Export wallet private key, encrypted with password
+    pub fn export_master_key(
+        &mut self,
+        session_id: types::SessionId,
+        wallet_id: String,
+        password: types::Password,
+    ) -> ResponseActFuture<String> {
+        let f = fut::result(
+            self.state
+                .get_wallet_by_session_and_id(&session_id, &wallet_id),
+        )
+        .and_then(move |wallet, slf: &mut Self, _| {
+            slf.params
+                .worker
+                .send(worker::ExportMasterKey(wallet, password))
+                .flatten()
+                .map_err(From::from)
+                .into_actor(slf)
+        });
+
+        Box::new(f)
+    }
+}
+
+// Validate `CreateWalletRequest`.
+///
+/// To be valid it must pass these checks:
+/// - password is at least 8 characters
+/// - seed_sources has to be `mnemonics | xprv`
+#[allow(clippy::too_many_arguments)]
+pub fn validate(
+    password: types::Password,
+    seed_data: types::Password,
+    seed_source: String,
+    name: Option<String>,
+    description: Option<String>,
+    overwrite: Option<bool>,
+    backup_password: Option<types::Password>,
+) -> Result<Validated> {
+    let source = match seed_source.as_ref() {
+        "xprv" => validate_xprv(seed_data, backup_password)
+            .map_err(|e| app::field_error("seed_data", e.to_string())),
+        "mnemonics" => Mnemonic::from_phrase(seed_data)
+            .map_err(|err| app::field_error("seed_data", format!("{}", err)))
+            .map(types::SeedSource::Mnemonics),
+        _ => Err(app::field_error(
+            "seed_source",
+            "Seed source has to be mnemonics|xprv",
+        )),
+    };
+    let password = if <str>::len(password.as_ref()) < 8 {
+        Err(app::field_error(
+            "password",
+            "Password must be at least 8 characters",
+        ))
+    } else {
+        Ok(password)
+    };
+    let overwrite = overwrite.unwrap_or(false);
+    app::combine_field_errors(source, password, move |seed_source, password| Validated {
+        description,
+        name,
+        overwrite,
+        password,
+        seed_source,
+    })
+    .map_err(validation_error)
+}
+
+/// Validate an encrypted XPRV file, first decrypting it and then checking the key format
+/// The seed data contains the hrp||iv||salt||ciphertext
+/// hrp can be either 'xprv' or 'xprvoduble'
+pub fn validate_xprv(
+    seed_data: types::Password,
+    backup_password: Option<types::Password>,
+) -> Result<types::SeedSource> {
+    let backup_password = backup_password.ok_or_else(|| {
+        validation_error(app::field_error(
+            "backup_password",
+            "Backup password not found for XPRV key",
+        ))
+    })?;
+    let seed_data_string = std::str::from_utf8(seed_data.as_ref()).map_err(|_| {
```

Comment:
This can only happen if seed_data is not valid UTF8, which I think is impossible because it is a `types::Password` which is just a type alias for `ProtectedString`, and `ProtectedString` must be valid UTF8.

But `ProtectedString` implements `AsRef<str>`, so you should be able to just do:

```
let seed_data_string: &str = seed_data.as_ref();
```

Reply:
great!